### notice

내일 오후에 자리로 오시면 안내해드릴게요!
- 일단 한 번 보시고 오세요~
- tips : 작동 원리가 잘 이해가 안 갈 때는 코드 사이 사이에 print를 넣어서 눈으로 확인해주는 법도 좋습니다.

In [7]:
class Calculator:
    def __init__(self):
        self.value = 0

    def add(self, val):
        self.value += val

In [2]:
# class MaxLimitCalculator(Calculator):
#     def add(self, val):
#         self.value += val
#         if self.value > 100:
#             self.value = 100

In [8]:
# 작성하신 것에 print문을 넣어보았습니다.

class MaxLimitCalculator(Calculator):
    def add(self, val):
        print(f'입력 된 값은 : {val}')
        if self.value >= 100:
            print('100이 넘어서 이쪽이 출력 되고 있습니다.')
            return 100
        else:
            print('else 쪽이 출력 되고 있습니다.')
            print(f'self.value 값은 : {self.value}')
            return self.value

In [9]:
cal = MaxLimitCalculator()
cal.add(50)  # 50 더하기
cal.add(60)  # 60 더하기

print(cal.value)  # 100 출력 

입력 된 값은 : 50
else 쪽이 출력 되고 있습니다.
self.value 값은 : 0
입력 된 값은 : 60
else 쪽이 출력 되고 있습니다.
self.value 값은 : 0
0


작성하신 코드를 살펴보겠습니다.
- cal = MaxLimitCalculator()를 생성합니다.
- cal.add(50) 실행
  - 코드 내에서 메소드 오버라이딩 때문에 self.value=0인 상태입니다.
  - add 메소드가 재정의 되었기 때문에 MaxLimitCalculator() 내에 있는 메소드를 사용합니다.
  - val = 50이 들어갑니다.
  - self.value는 여전히 0이기 때문에 else 구문으로 이동합니다.
  - self.value에 더해지는 값이 없기 때문에 계속 0이 됩니다.
  - 빠뜨린 부분은 self.value += val이겠습니다.
  - 물론 이렇게 한다고 문제에서 요구하는 것이 다 반영된 건 아닙니다.
  
- cal.add(60) 실행
  - 마찬가지 결과로 self.value에는 더해지는 값이 없기 때문에 0입니다.

- print(cal.value)
  - 최종적으로 아무것도 더해진 게 없으므로 cal.value는 0 입니다.

In [10]:
class MaxLimitCalculator(Calculator):
    def add(self, val):
        print(f'입력 된 값은 : {val}')
        self.value += val
        print(f'현재 self.value 값은 : {self.value}')
        
        if self.value >= 100:
            print('100이 넘어서 이쪽이 출력 되고 있습니다.')
            self.value = 100
        else:
            print('else 쪽이 출력 되고 있습니다.')
            print(f'self.value : {self.value}')

In [11]:
cal = MaxLimitCalculator()
cal.add(50)  # 50 더하기
cal.add(60)  # 60 더하기

print(cal.value)  # 100 출력 

입력 된 값은 : 50
현재 self.value 값은 : 50
else 쪽이 출력 되고 있습니다.
self.value : 50
입력 된 값은 : 60
현재 self.value 값은 : 110
100이 넘어서 이쪽이 출력 되고 있습니다.
100


위와 다른 점은 인스턴스가 생성될 때 만들어지는 값인 self.value에 val 값을 우선 더하는 것 입니다.
- 코드는 순차적으로 실행 되기 때문에 먼저 값을 정해주고 가야 합니다.


입력 된 값이 50
- 현재 self.value 값은 50

입력 된 값이 60
- self.value += val이 있기 때문에 두 개의 합은 110이 됩니다.
- 그리고 if문에서 self.value를 판단하고 값을 100으로 넘겨줍니다.
- 그래서 마지막에 100이 나오게 됩니다.
- 문제의 해답과 같은 원리입니다.